hashes/src/util.rs

```diff
@@ -15,33 +15,41 @@
 #[macro_export]
 /// Adds hexadecimal formatting implementation of a trait `$imp` to a given type `$ty`.
 macro_rules! hex_fmt_impl(
-    ($ty:ident) => (
-        $crate::hex_fmt_impl!($ty, );
+    ($reverse:expr, $ty:ident) => (
+        $crate::hex_fmt_impl!($reverse, $ty, );
     );
-    ($ty:ident, $($gen:ident: $gent:ident),*) => (
+    ($reverse:expr, $ty:ident, $($gen:ident: $gent:ident),*) => (
         impl<$($gen: $gent),*> $crate::_export::_core::fmt::LowerHex for $ty<$($gen),*> {
+            #[inline]
             fn fmt(&self, f: &mut $crate::_export::_core::fmt::Formatter) -> $crate::_export::_core::fmt::Result {
-                #[allow(unused_imports)]
-                use $crate::{Hash as _, HashEngine as _, hex};
-
-                if f.alternate() {
-                    write!(f, "0x")?;
+                if $reverse {
+                    $crate::_export::_core::fmt::LowerHex::fmt(&self.0.backward_hex(), f)
+                } else {
+                    $crate::_export::_core::fmt::LowerHex::fmt(&self.0.forward_hex(), f)
                 }
-                if $ty::<$($gen),*>::DISPLAY_BACKWARD {
-                    hex::format_hex_reverse(self.as_ref(), f)
+            }
+        }
+
+        impl<$($gen: $gent),*> $crate::_export::_core::fmt::UpperHex for $ty<$($gen),*> {
+            #[inline]
+            fn fmt(&self, f: &mut $crate::_export::_core::fmt::Formatter) -> $crate::_export::_core::fmt::Result {
+                if $reverse {
+                    $crate::_export::_core::fmt::UpperHex::fmt(&self.0.backward_hex(), f)
                 } else {
-                    hex::format_hex(self.as_ref(), f)
+                    $crate::_export::_core::fmt::UpperHex::fmt(&self.0.forward_hex(), f)
                 }
             }
         }
 
         impl<$($gen: $gent),*> $crate::_export::_core::fmt::Display for $ty<$($gen),*> {
+            #[inline]
             fn fmt(&self, f: &mut $crate::_export::_core::fmt::Formatter) -> $crate::_export::_core::fmt::Result {
-                $crate::_export::_core::fmt::LowerHex::fmt(self, f)
+                $crate::_export::_core::fmt::LowerHex::fmt(&self.0, f)
```

Comment:
This should be `self` not `self.0` because with `self.0` we ignore `$reverse`.

Reply:
Generics have to be at the end and I think it'd be a bit weird to have `$reverse` between type name and generics. Agree this one is not great either. :man_shrugging:

Reply:
This one doesn't
```rust

    #[test]
    fn hash_newtype_reverse() {
        // sha256::Hash is displayed forwards.
        hash_newtype!(TestHash, sha256::Hash, 32, doc="test hash type that should be displayed backwards", true);

        let inner = sha256::Hash::hash(&[0_u8; 32]);
        let h = TestHash::from_hash(inner);

        let display = format!("{}", h);
        let lower_hex = format!("{:x}", h);

        assert_eq!(display, lower_hex);
    }
```

Reply:
Ah, great, will fix after #1477 is in.